### PR TITLE
watchtower: start using the DeleteSession message

### PR DESCRIPTION
**Pull request by ellemouton**
In this PR, we start making use of of the `DeleteSession` message. This message allows a tower client to indicate to a tower server that it can delete sessions. Sessions are considered "deletable" if they are exhausted & if all the acked-updates sent on the session are for channels that are closed. 

This allows us to claim back disk space on both the client and server.

Fixes https://github.com/lightningnetwork/lnd/issues/6259

Related: https://github.com/lightningnetwork/lnd/issues/7121

# Main Changes:

## 1. Add a new Session-ID index
A new session-ID index is added. The index holds a mapping from db-assigned-ID (BigSize) to realsession-ID (33 bytes). This mapping will allow us to preserve disk space in future when persisting references to sessions.

## 2. Add a channel -> session ID index:
This is added so that it is quick to check which sessions should be checked if a channel is closed.

## 3. Channel Close handler:
The tower client now subscribes to channel closures. Once a notification is received, it is marked as closed in the tower DB along with the close block height. We then check each of the session that have acked updates for the tower to see if they are closable - if they are, then we add the sessionID to the new `closable sessions` bucket. With this change, we can now also stop loading the summaries of closed channels on startup. 

## 4. Closable Sessions handler:
We use a min-heap to store all the closable-sessions.
The tower client now subscribes to new blocks notifications. On each new block, we check the min-heap to see if there are sessions we should close. If there are, then we dial the tower and send the `DeleteSession` message and we also delete all the session info from our DB.

### CODE REVIEW

File: watchtower/wtdb/client_db.go [hunk 1]
```diff
@@ -959,6 +1373,372 @@ func (c *ClientDB) MarkBackupIneligible(chanID lnwire.ChannelID,
...
+		return nil, err
+	}
+
+	return closableSessions, nil
+}
+
+// isSessionClosable returns true if a session is considered closable. A session
+// is considered closable only if: 1) It has no un-acked updates, 2) It is
+// exhausted (ie it cant accept any more updates), 3) All the channels that it
+// has acked upates for are all closed.
```
Comment (C-Otto) [inline watchtower/wtdb/client_db.go:1662]:
```suggestion
// has acked updates for are closed.
```

File: watchtower/wtdb/client_db.go [hunk 2]
```diff
@@ -959,6 +1373,372 @@ func (c *ClientDB) MarkBackupIneligible(chanID lnwire.ChannelID,
...
+		chanID, err := getRealChannelID(
+			chanIDIndexBkt, byteOrder.Uint64(dbChanID),
+		)
+		if err != nil {
+			return err
+		}
+
+		chanDetails := chanDetailsBkt.NestedReadBucket(chanID[:])
+		if chanDetails == nil {
+			return nil
```
Comment (C-Otto) [inline watchtower/wtdb/client_db.go:1719]: Under which conditions can this happen? Would it be possible for some very old channel to be unregistered, but still open? In other words, how large is the risk of assuming an unknown channel is closed?

File: watchtower/wtclient/client.go [hunk 3]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+
+	for {
+		select {
+		case newBlock := <-blocksChan.Epochs:
+			if newBlock == nil {
+				return
+			}
+
+			height := uint32(newBlock.Height)
+			for {
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:970]: As far as I can see, this would delete all eligible old sessions in one go. I currently have ~50k sessions, and I guess that >50% of those can be deleted. However, I also think that doing 50k deletions in this loop would cause some issues both for my node (WT client) and my WT server(s). Would it be OK to limit this to, say, 1k sessions per block?

File: watchtower/wtclient/client.go [hunk 4]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+				sess, err := c.cfg.DB.GetClientSession(
+					item.sessionID,
+				)
+				if err != nil {
+					log.Error("error calling "+
+						"GetClientSession: %v", err)
+					continue
+				}
+
+				err = c.deleteSessionFromTower(sess)
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:999]: I guess this can happen if the tower is offline? If so, would the deletion be tried again? I think this would only happen once lnd (WT client) is restarted, which doesn't seem too bad.

File: watchtower/wtclient/client.go [hunk 5]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+					chanID, closeInfo.CloseHeight,
+				)
+				if err != nil {
+					log.Errorf("could not handle closed "+
+						"channel(%s): %v", chanID, err)
+					continue
+				}
+
+				for _, sID := range sl {
+					delay, err := newRandomDelay(
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:912]: "for given session ID and block height, compute random delay and push into closableSessionQueue" is duplicated (line 618)

File: watchtower/wtclient/client.go [hunk 6]
```diff
@@ -545,10 +570,72 @@ func (c *TowerClient) Start() error {
...
+
+		// Load all closable sessions.
+		sl, err := c.cfg.DB.ListClosableSessions()
+		if err != nil {
+			returnErr = err
+			return
+		}
+
+		for sID, blockHeight := range sl {
+			delay, err := newRandomDelay(c.cfg.SessionCloseRange)
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:610]: Duplication, see line 912

File: watchtower/wtclient/client.go [hunk 7]
```diff
@@ -545,10 +570,72 @@ func (c *TowerClient) Start() error {
...
+				log.Errorf("could not handle closed "+
+					"channel(%s): %v", id, err)
+			}
+		}
+
+		c.wg.Add(1)
+		go c.handleChannelCloses(chanSub)
+
+		// Load all closable sessions.
+		sl, err := c.cfg.DB.ListClosableSessions()
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:603]: Nit: `sl` doesn't really tell a story, I'd prefer `closableSessions`

File: watchtower/wtclient/client.go [hunk 8]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+			if iteratorErr == nil {
+				towerAddr = nextAddr
+				continue
+			}
+
+			// Otherwise, if we have exhausted the address list,
+			// exit.
+			tower.Addresses.Reset()
+
+			return fmt.Errorf("failed to dial tower(%s) at any "+
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:1078]: This doesn't look right:
```
[ERR] WTCL: error deleting session from tower: failed to dial tower(^b?ڢ???^|;'t."??6??^^o??1?t any available addresses
```

**Comment by C-Otto** [comment]
Alright, it works! :partying_face: The number of sessions reported by `lncli wtclient towers` is slowly going down, and according to some added log statements the sessions are deleted both from the remote towers and from the local DB. It takes around 7 seconds per session, though - even though I moved my slow tower to the faster SSD storage. Thanks a lot, @ellemouton!

File: watchtower/wtclient/client.go [hunk 9]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+				c.closableSessionQueue.Pop()
+
+				// Fetch the session from the DB so that we can
+				// extract the Tower info.
+				sess, err := c.cfg.DB.GetClientSession(
+					item.sessionID,
+				)
+				if err != nil {
+					log.Error("error calling "+
+						"GetClientSession: %v", err)
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:995]: Please also include some context for errors, in this case the session ID would help.

File: watchtower/wtclient/client.go [hunk 10]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+					continue
+				}
+
+				err = c.deleteSessionFromTower(sess)
+				if err != nil {
+					log.Error("error deleting session "+
```
[the remark this answers is not in the document]
Comment (C-Otto) [reply watchtower/wtclient/client.go:1001]: As before, please also add context (session ID).

File: watchtower/wtclient/client.go [hunk 11]
```diff
@@ -777,6 +865,277 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+						"from tower: %v", err)
+					continue
+				}
+
+				err = c.cfg.DB.DeleteSession(item.sessionID)
+				if err != nil {
+					log.Errorf("could not delete "+
+						"session(%s) from DB: %v", err)
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:1009]: The value for `%s` is missing? Should it be `%v`?

**Comment by C-Otto** [comment]
I also get a bunch of these:
```
[ERR] WTCL: error calling GetClientSession for 032b5e4e40ea7fbf33bc284b05082675de9dffa2070625a90b4050aef689be505b: client session not found
```

(I tweaked the log message)

File: watchtower/wtclient/client.go [hunk 12]
```diff
@@ -545,10 +570,72 @@ func (c *TowerClient) Start() error {
...
 
+		chanSub, err := c.cfg.SubscribeChannelEvents()
+		if err != nil {
+			returnErr = err
+			return
+		}
+
+		// Iterate over the list of registered channels and check if
+		// any of them can be marked as closed.
+		for id := range c.summaries {
```
Comment (C-Otto) [inline watchtower/wtclient/client.go:581]: The code initializing the summaries uses `cfg.DB.FetchChanSummaries()`, which does not return closed summaries. On my node, I was able to delete ~20% of all sessions. I don't know why it stopped marking other sessions as deletable, but I assume that I'd suffice to re-evaluate based on the existing code and `c.summaries` including closed channels (!).

**Comment by C-Otto** [comment]
lnd managed to delete around 20% of all sessions, which (aside from a few errors I hinted at above) worked without any issues. The remaining 80% cannot be deleted, yet, as they are referenced by at least one open channel each. Based on a quick analysis, I have a single channel that blocks 3100 sessions. Another channel blocks 2200 sessions. Both of these channels are rather old and active, which reinforces my trust in the code.

To summarize: as far as I can see everything works as expected, and I'm confident these older sessions will disappear once the remaining channels are gone, too. Thanks a lot!

**Review by ellemouton — COMMENTED** [review]
thanks for your review & for testing this @C-Otto !

I have addressed most of your comments & also added some unit tests. 
Last TODO is to add an itest for this. 

I haven't yet been able to determine why you have run into those error logs.... hopefully I will be able to reproduce those in either the itest or in a stress test.

**Comment by C-Otto** [comment]
> I haven't yet been able to determine why you have run into those error logs.... hopefully I will be able to reproduce those in either the itest or in a stress test.

The WT server is rather slow and didn't respond in time, even though it might have deleted some session. It's likely that these errors popped up after a restart when the client tried to delete these sessions again.

**Comment by C-Otto** [comment]
```
2022-12-05 11:45:26.104 [ERR] WTCL: error deleting session(0255d75aa2039735d8480b771eddac7dcb07b3b406574707df2a3039727059f315) from tower: failed to dial tower([2 131 [...]]) at any available addresses
2022-12-05 11:45:26.108 [ERR] WTCL: error deleting session(0255d75aa2039735d8480b771eddac7dcb07b3b406574707df2a3039727059f315) from tower: failed to dial tower([2 131 [...]]) at any available addresses
2022-12-05 14:00:40.950 [ERR] WTCL: error calling GetClientSession for 0255d75aa2039735d8480b771eddac7dcb07b3b406574707df2a3039727059f315: client session not found
```

I restarted lnd at 13:53, so it seems "error calling GetClientSession" corresponds to a successful deletion on the server where the client didn't get the response in time.

**Comment by C-Otto** [comment]
I don't have any recent log entry for "could not handle closed channel", maybe this is just related to my initial experimentation. The only entries are from a 5 minute period on November 28th, so feel free to ignore this.

File: sample-lnd.conf [hunk 13]
```diff
@@ -991,6 +991,13 @@ litecoin.node=ltcd
 ; supported at this time, if none are provided the tower will not be enabled.
 ; wtclient.private-tower-uris=
 
+; The range over which to choose a random number of block to wait after the
```
Comment (C-Otto) [inline sample-lnd.conf:994]:
```suggestion
; The range over which to choose a random number of blocks to wait after the
```

File: watchtower/wtdb/client_db.go [hunk 14]
```diff
@@ -1378,6 +1378,44 @@ func (c *ClientDB) MarkBackupIneligible(chanID lnwire.ChannelID,
...
+			sessID, err := getRealSessionID(sessIDIndexBkt, dbID)
+			if err != nil {
+				return err
+			}
+
+			sessions[*sessID] = byteOrder.Uint32(heightBytes)
+
+			return nil
+		})
+	}, func() {})
```
Comment (bhandras) [inline watchtower/wtdb/client_db.go:1429]: nit: isn't really an issue in this case but I'd still reset `sessions` here.

File: watchtower/wtdb/client_db.go [hunk 15]
```diff
@@ -1416,6 +1426,177 @@ func (c *ClientDB) ListClosableSessions() (map[SessionID]uint32, error) {
...
+			return err
+		}
+
+		// Delete from the tower-to-sessionID index.
+		towerIndexBkt := towerToSessBkt.NestedReadWriteBucket(
+			sess.TowerID.Bytes(),
+		)
+		if towerIndexBkt == nil {
+			return fmt.Errorf("no entry in the tower-to-session "+
+				"index found for tower ID %x", sess.TowerID)
```
Comment (bhandras) [inline watchtower/wtdb/client_db.go:1502]: nit: do we want to `%v` instead?

File: watchtower/wtclient/candidate_iterator_test.go [hunk 16]
```diff
@@ -155,4 +161,16 @@ func TestTowerCandidateIterator(t *testing.T) {
 	towerIterator.AddCandidate(secondTower)
 	assertActiveCandidate(t, towerIterator, secondTower, true)
 	assertNextCandidate(t, towerIterator, secondTower)
+
+	// Assert that the GetTower correctly returns the tower too.
+	tower, err := towerIterator.GetTower(secondTower.ID)
```
Comment (bhandras) [inline watchtower/wtclient/candidate_iterator_test.go:166]: :+1:

File: watchtower/wtclient/interface.go [hunk 17]
```diff
@@ -197,3 +197,30 @@ type ClientSession struct {
 	// key used to connect to the watchtower.
 	SessionKeyECDH keychain.SingleKeyECDH
 }
+
+// NewClientSessionFromDBSession converts a wtdb.ClientSession to a
+// ClientSession.
+func NewClientSessionFromDBSession(s *wtdb.ClientSession, tower *Tower,
```
Comment (bhandras) [inline watchtower/wtclient/interface.go:204]: :taco:

File: watchtower/wtclient/client.go [hunk 18]
```diff
@@ -533,10 +546,44 @@ func (c *TowerClient) Start() error {
...
+				return
+			}
+
+			if !isClosed {
+				continue
+			}
+
+			_, err = c.cfg.DB.MarkChannelClosed(id, closedHeight)
+			if err != nil {
+				log.Errorf("could not handle closed "+
```
Comment (bhandras) [inline watchtower/wtclient/client.go:590]: nit: wdyt about "could not mark channel(%s) as closed: %v"

File: watchtower/wtclient/client.go [hunk 19]
```diff
@@ -765,6 +813,66 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
 }
 
+// handleChannelCloses listens for channel close events and marks channels as
+// closed in the DB.
+//
+// NOTE: This method MUST be run as a goroutine.
+func (c *TowerClient) handleChannelCloses(chanSub subscribe.Subscription) {
+	defer c.wg.Done()
+
+	c.log.Tracef("Starting channel close handler")
```
Comment (bhandras) [inline watchtower/wtclient/client.go:868]: nit: might as well be `Debugf` imo.

File: watchtower/wtclient/client.go [hunk 20]
```diff
@@ -765,6 +813,66 @@ func (c *TowerClient) nextSessionQueue() (*sessionQueue, error) {
...
+			}
+
+			chanID := lnwire.NewChanIDFromOutPoint(
+				&event.CloseSummary.ChanPoint,
+			)
+
+			// We only care about channels registered with the tower
+			// client.
+			c.backupMu.Lock()
+			_, ok = c.summaries[chanID]
```
Comment (bhandras) [inline watchtower/wtclient/client.go:893]: Q: could there be a race condition where we add the channel to the summaries later than we receive a close notification?

File: watchtower/wtclient/client_test.go [hunk 21]
```diff
@@ -1654,6 +1754,142 @@ var clientTests = []clientTest{
...
+			},
+		},
+		fn: func(h *testHarness) {
+			const numUpdates = 5
+
+			// In this test we assert that a channel is correctly
+			// marked as closed and that sessions are also correctly
+			// marked as closable.
+
+			// We start with the sendUpdatesOn parameter set to
```
Comment (bhandras) [inline watchtower/wtclient/client_test.go:1829]: Love these detailed comments!

File: watchtower/wtclient/sess_close_min_heap.go [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+type sessionCloseItem struct {
+	// sessionID is the ID of the session in question.
+	sessionID wtdb.SessionID
+
+	// deleteHeight is the block height after which we can delete the
+	// session.
+	deleteHeight uint32
+}
+
+// Less returns true if the current item's delete height is larger than the
```
Comment (bhandras) [inline watchtower/wtclient/sess_close_min_heap.go:84]: nit: larger => less

File: lncfg/wtclient.go [hunk 23]
```diff
@@ -17,6 +17,11 @@ type WtClient struct {
 	// SweepFeeRate specifies the fee rate in sat/byte to be used when
 	// constructing justice transactions sent to the tower.
 	SweepFeeRate uint64 `long:"sweep-fee-rate" description:"Specifies the fee rate in sat/byte to be used when constructing justice transactions sent to the watchtower."`
+
+	// SessionCloseRange is the range over which to choose a random number
+	// of blocks to wait after the last channel of a session is closed
+	// before sending the DeleteSession message to the tower server.
+	SessionCloseRange uint32 `long:"session-close-range" description:"The range over which to choose a random number of blocks to wait after the last channel of a session is closed before sending the DeleteSession message to the tower server. Set to 1 for no delay."`
```
Comment (bhandras) [inline lncfg/wtclient.go:24]: This is to be reorg safe?

File: watchtower/wtclient/client.go [hunk 24]
```diff
@@ -876,6 +877,125 @@ func (c *TowerClient) handleChannelCloses(chanSub subscribe.Subscription) {
...
+		conn, err = c.dial(
+			session.SessionKeyECDH, &lnwire.NetAddress{
+				IdentityKey: tower.IdentityKey,
+				Address:     towerAddr,
+			},
+		)
+		if err != nil {
+			// If there are more addrs available, immediately try
+			// those.
+			nextAddr, iteratorErr := tower.Addresses.Next()
```
Comment (bhandras) [inline watchtower/wtclient/client.go:1098]: Q: are we sure the iterator isn't advanced/reset elsewhere?

File: lntest/itest/lnd_watchtower_test.go [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+		var (
+			numBackups uint32
+		)
+		for _, session := range info.Sessions {
+			numBackups += session.NumBackups
+		}
+
+		return numBackups == 0
+	}, defaultTimeout)
+	require.NoError(t.t, err)
```
Comment (bhandras) [inline lntest/itest/lnd_watchtower_test.go:190]: Very nice itest!!

**Review by bhandras — APPROVED** [review]
Super clean and readable PR! Awesome work @ellemouton :1st_place_medal: 
LGTM :milk_glass:

File: watchtower/wtdb/client_db.go [hunk 26]
```diff
@@ -719,24 +729,58 @@ func (c *ClientDB) CreateClientSession(session *ClientSession) error {
...
 		if err != nil {
 			return err
 		}
 
+		err = sessionBkt.Put(cSessionDBID, newIndex)
+		if err != nil {
+			return err
+		}
+
+		// TODO(elle): migrate the towerID-to-SessionID to use the
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go:802]: yes :) I dont think the storage improvements are (at least at the moment) significant enough to do the migration in this PR. But could be worth it later on. So just left the todo so that it is clear that we could possibly get some storage gains here if needed.

File: watchtower/wtdb/client_db.go [hunk 27]
```diff
@@ -719,24 +729,58 @@ func (c *ClientDB) CreateClientSession(session *ClientSession) error {
...
+		// new db-assigned sessionID's rather.
+
+		// Add the new entry to the towerID-to-SessionID index.
+		towerToSessIndex := towerToSessionIndex.NestedReadWriteBucket(
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: done!

File: watchtower/wtdb/client_db.go [hunk 28]
```diff
@@ -1859,6 +1903,30 @@ func getDBChanID(chanDetailsBkt kvdb.RBucket, chanID lnwire.ChannelID) (uint64,
...
 }
 
+// getDBSessionID returns the db-assigned session ID for the given real session
+// ID. It returns both the uint64 and byte representation.
+func getDBSessionID(sessionsBkt kvdb.RBucket, sessionID SessionID) (uint64,
+	[]byte, error) {
+
+	sessionBkt := sessionsBkt.NestedReadBucket(sessionID[:])
+	if sessionBkt == nil {
+		return 0, nil, ErrChannelNotRegistered
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: indeed! thanks

File: watchtower/wtdb/migration5/client_db.go [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+	// Iterate over the sessions bucket where each key is a session-ID.
+	return sessionsBkt.ForEach(func(sessionID, _ []byte) error {
+		// Ask the DB for a new, unique, id for the index bucket.
+		nextSeq, err := indexBkt.NextSequence()
+		if err != nil {
+			return err
+		}
+
+		var newIndex bytes.Buffer
+		err = tlv.WriteVarInt(&newIndex, nextSeq, &[8]byte{})
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/migration5/client_db.go]: done!

File: watchtower/wtdb/client_db.go [hunk 30]
```diff
@@ -1005,6 +1005,39 @@ func getSessionKeyIndex(keyIndexes kvdb.RwBucket, towerID TowerID,
...
+			sessionsBkt, chanIDIndexBkt, id[:], opts...,
+		)
+		if err != nil {
+			return err
+		}
+
+		sess = session
+		return nil
+	}, func() {})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: true story. Done!

File: watchtower/wtdb/client_db.go [hunk 31]
```diff
@@ -1358,6 +1378,208 @@ func (c *ClientDB) MarkBackupIneligible(chanID lnwire.ChannelID,
...
+			if err != nil {
+				return err
+			}
+
+			closableSessions = append(closableSessions, *sID)
+
+			return nil
+		})
+	}, func() {
+		closableSessions = []SessionID{}
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: doooone

File: watchtower/wtdb/client_db.go [hunk 32]
```diff
@@ -1358,6 +1378,208 @@ func (c *ClientDB) MarkBackupIneligible(chanID lnwire.ChannelID,
...
+	}
+
+	session, err := getClientSessionBody(sessionsBkt, id[:])
+	if err != nil {
+		return false, err
+	}
+
+	// We have already checked that the session has no more committed
+	// updates. So now we can check if the session is exhausted.
+	if session.SeqNum != session.Policy.MaxUpdates {
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: good point!

File: watchtower/wtdb/client_db.go (hunk 15 above)
Comment (ellemouton) [reply watchtower/wtdb/client_db.go]: good catch!

File: watchtower/wtclient/client.go (hunk 20 above)
Comment (ellemouton) [reply watchtower/wtclient/client.go:894]: even if that is possible, I dont think it would be an issue: if we first get a close-notification & then only register the channel then it is still the case that the channel is in fact closed. So no updates will be backed up for that channel. So no sessions would be affected by that channel closure. Do you agree?

File: watchtower/wtclient/sess_close_min_heap.go (hunk 22 above)
Comment (ellemouton) [reply watchtower/wtclient/sess_close_min_heap.go]: 🤓

File: watchtower/wtclient/client.go (hunk 24 above)
Comment (ellemouton) [reply watchtower/wtclient/client.go:1098]: yeah it is. My thinking about this is that in most cases, the tower will have 1 address that just works and this will be seen by all threads using the tower iterator when they call `Peak()`. Then, since all the threads using these iterators are best-effort, they will eventually try again at some point if the Peak address does in fact fail. 

What I could do though is add a `Copy` method to the address iterator so that this method acts on a completely independent iterator. The only thing that that will then mean is that any addresses added while this method is running wont be seen. But again, since it is best effort and will try again on restart, that isnt such a big deal. 

Thoughts?

File: lncfg/wtclient.go (hunk 23 above)
Comment (ellemouton) [reply lncfg/wtclient.go:24]: discussed offline but mentioning here for other reviewers:

It is not about re-org safety but is rather for privacy. We dont necessarily want to tell the tower that it can delete a session _as soon as_ all the associated channels are closed cause then it will be able to do some chain-analysis things to figure out which channels you own. So this parameter lets the user choose a range from which a numbe-number-of-blocks-to-wait-after-channel-closure will be randomly chosen.

**Review by ellemouton — COMMENTED** [review]
Thanks for the review @bhandras !

File: watchtower/wtdb/client_db.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (bhandras) [reply watchtower/wtdb/client_db.go:802]: sgtm

File: watchtower/wtclient/client.go (hunk 20 above)
Comment (bhandras) [reply watchtower/wtclient/client.go:894]: Agree and sounds good!

File: watchtower/wtclient/client.go (hunk 24 above)
Comment (ellemouton) [reply watchtower/wtclient/client.go]: added! 🚀

File: watchtower/wtdb/client_db.go [hunk 33]
```diff
@@ -1993,6 +2215,44 @@ func getDBSessionID(sessionsBkt kvdb.RBucket, sessionID SessionID) (uint64,
...
+		return nil, fmt.Errorf("session ID not found")
+	}
+
+	var sessID SessionID
+	copy(sessID[:], sessIDBytes)
+
+	return &sessID, nil
+}
+
+func getRealChannelID(chanIDIndexBkt kvdb.RBucket, dbID uint64) (
```
Comment (positiveblue) [inline watchtower/wtdb/client_db.go:2466]: nit: func format

**Review by ziggie1984 — COMMENTED** [review]
Really good job bringing storage-efficiency to the wtclient 👍 

Did a quick pass through will look at it again today with more detail!

File: watchtower/wtclient/client.go [hunk 34]
```diff
@@ -878,6 +879,130 @@ func (c *TowerClient) handleChannelCloses(chanSub subscribe.Subscription) {
...
+
+	session, err := NewClientSessionFromDBSession(
+		sess, tower, c.cfg.SecretKeyRing,
+	)
+	if err != nil {
+		return err
+	}
+
+	localInit := wtwire.NewInitMessage(
+		lnwire.NewRawFeatureVector(wtwire.AltruistSessionsRequired),
```
Comment (ziggie1984) [inline watchtower/wtclient/client.go:1079]: Q: I wonder if you could make this more flexible to allow for Reward-Based towers, which are already part of the codebase ?

**Review by ziggie1984 — COMMENTED** [review]
Looked through it in detail, found some nits. All in all very good PR. Comments, Tests help a lot to understand the code.

File: itest/lnd_watchtower_test.go [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+	}, defaultTimeout)
+	require.NoError(ht, err)
+
+	// Now close the channel and wait for the close transaction to appear
+	// in the mempool so that it is included in a block when we mine.
+	ht.CloseChannelAssertPending(dave, chanPoint, true)
+
+	// Mine enough blocks to surpass the session-close-range. This should
+	// trigger the session to be deleted.
+	ht.MineBlocksAndAssertNumTxes(sessionCloseRange+5, 1)
```
Comment (positiveblue) [inline itest/lnd_watchtower_test.go:135]: not blocking: It would be nice to be able to check that the session is not deleted before the assigned `deleteHeight` but I guess it s too hard to test now because 1) we can not change it from here 2) we cannot even inspect it here. I am not sure if this specific check is performed in [client_test.go](https://github.com/lightningnetwork/lnd/pull/7069/commits/a2a93c291980236729e12685f3c41351bd4eea8d#diff-0958e6d1e15ea7d13d811653ec1d995207be37f60da366dcaeffacec9b5582b1R1947)

**Review by positiveblue — APPROVED** [review]
Amazing work @ellemouton 

Nothing blocking from my side :rocket:

File: itest/lnd_watchtower_test.go [hunk 36]
```diff
@@ -0,0 +1,162 @@
...
+			for _, session := range sessionType.Sessions {
+				numBackups += session.NumBackups
+			}
+		}
+
+		return numBackups == 0
+	}, defaultTimeout)
+	require.NoError(ht, err)
+
+	ht.CleanupForceClose(ht.Alice, chanPoint)
```
Comment (yyforyongyu) [inline itest/lnd_watchtower_test.go:161]: this needs to be `Dave`

File: itest/lnd_watchtower_test.go [hunk 37]
```diff
@@ -0,0 +1,162 @@
...
+	dave := ht.NewNode("Dave", daveArgs)
+
+	addTowerReq := &wtclientrpc.AddTowerRequest{
+		Pubkey:  wallisInfo.Pubkey,
+		Address: listener,
+	}
+	dave.RPC.AddTower(addTowerReq)
+
+	// Assert that there exists a session between Dave and Wallis.
+	err := wait.Predicate(func() bool {
```
Comment (yyforyongyu) [inline itest/lnd_watchtower_test.go:67]: had a discussion with @guggero a while back, I think we want to unify the usage and stick to `wait.NoError` only as it provides more verbose error log.

File: itest/lnd_watchtower_test.go [hunk 38]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/lightningnetwork/lnd/lntest/wait"
+	"github.com/stretchr/testify/require"
+)
+
+// testWatchtowerSessionManagement tests that session deletion is done
+// correctly.
+func testWatchtowerSessionManagement(ht *lntest.HarnessTest) {
+	const (
+		chanAmt           = funding.MaxBtcFundingAmount
+		paymentAmt        = 100
```
Comment (yyforyongyu) [inline itest/lnd_watchtower_test.go:22]: you may wanna increase this value to `10_000` to avoid the error,
```
2023-03-17 19:33:44.755 [DBG] SWPR: 1 negative yield inputs not added to input set: f8719e811940e354a9fb964430fd9236bfcb9bc663e855fd1b8fd6159c80440e:0 (CommitmentNoDelayTweakless)
```
guess sth is up with our sweeper🧐

**Review by ziggie1984 — APPROVED** [review]
Thanks for incorporating my comments, awesome job will orientate my future PRs especially on your commit structure ;)

**Comment by guggero** [comment]
@ellemouton looks like one of the commits doesn't compile. And the itests seem to fail on a watchtower test.

**Comment by ellemouton** [comment]
Yes - it started giving problems after rebasing. Still trying to figure out why the channel is nit being cleaned up by the test. Will ping you once it is ready

File: watchtower/wtclient/sess_close_min_heap.go [hunk 39]
```diff
@@ -0,0 +1,95 @@
+package wtclient
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply watchtower/wtclient/sess_close_min_heap.go:1]: opting to keep as is for now cause the `lnd/queue` package is a module so would need another PR, tagging etc and not sure if tags would transfer to `master` from `0.16.1-staging`.

File: itest/lnd_watchtower_test.go (hunk 35 above)
Comment (ellemouton) [reply itest/lnd_watchtower_test.go:158]: yeah - a bit tricky

File: itest/lnd_watchtower_test.go (hunk 37 above)
Comment (ellemouton) [reply itest/lnd_watchtower_test.go]: changed to `NoError` 👍

File: itest/lnd_watchtower_test.go (hunk 38 above)
Comment (ellemouton) [reply itest/lnd_watchtower_test.go]: cool - managed to get things working by just mining a few blocks at the end of the test to give both sweepers a chance to sweep

**Comment by ellemouton** [comment]
@C-Otto  - I think I found & fixed the main cause for the various errors you were seeing 🎉
